File: app/src/main/java/com/motrixi/datacollection/service/MotrixiService.java
package com.motrixi.datacollection.service;

import android.app.Notification;
import android.app.NotificationChannel;
import android.app.NotificationManager;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.os.Build;
import android.os.IBinder;
import android.support.annotation.Nullable;
import android.support.annotation.RequiresApi;
import android.util.Log;

import com.motrixi.datacollection.content.Contants;
import com.motrixi.datacollection.content.Session;
import com.motrixi.datacollection.utils.UploadCollectedData;
import com.motrixi.datacollection.utils.UploadLogUtil;

import java.util.Date;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledThreadPoolExecutor;
import java.util.concurrent.TimeUnit;

import static com.motrixi.datacollection.content.Contants.lastSyncTime;

/**
 * author : Jason
 * date   : 2020/12/8 4:08 PM
 * desc   :
 */
public class MotrixiService extends Service {

    private static ScheduledExecutorService startUploadExecutor;
    private static String CHANNEL_ID = "motrixi_sdk_channel";
    private static String CHANNEL_NAME = "motrixi_sdk_foreground_service";
    private int id = 100;
    //    private val TIME_VALUE: Long = 24*60*60*1000
    private static Long TIME_VALUE = Long.valueOf(10*60*1000);

    @Override
    public void onCreate() {
        super.onCreate();
        Log.e("oncreate","create service");

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            setForeground();
        }
    }

    @Nullable
    @Override
    public IBinder onBind(Intent intent) {
        return null;
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {

        startUploading();
        return super.onStartCommand(intent, flags, startId);
    }

    @RequiresApi(Build.VERSION_CODES.O)
    private void setForeground() {
        NotificationManager manager = (NotificationManager) getSystemService(Context.NOTIFICATION_SERVICE);
        NotificationChannel channel = new NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_HIGH);
        manager.createNotificationChannel(channel);
        Notification notification = new Notification.Builder(this, CHANNEL_ID)
                //.setContentTitle("收到一条重要通知")
                //.setContentText("这是重要通知")
                //.setSmallIcon(R.mipmap.ic_launcher)
//            .setLargeIcon(BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher))
                .build();
        startForeground(id, notification);
        Log.e("action","startForeground service");
    }

    /**
     * start upload the data
     */
    private void startUploading() {

        try {
            if (startUploadExecutor == null) {
                /*startUploadExecutor = ScheduledThreadPoolExecutor(1,
                    BasicThreadFactory.Builder().namingPattern("start-thread").daemon(true).build())*/
                startUploadExecutor = new ScheduledThreadPoolExecutor(1);
            }
            startUploadExecutor.scheduleAtFixedRate(new Runnable(){
                @Override
                public void run() {
                    long currentTime = new Date().getTime();
                    long lastTime = lastSyncTime;
                    if (lastTime == 0) {
                        Log.d("not confirm", "have not confirm");
                        return;
                    }
                    if ((currentTime - lastTime) >= TIME_VALUE) {

                        Log.d("service", "start service");
                        //UploadLogUtil.uploadLogData(Contants.mFREContext, "uploading data");
                        lastSyncTime = new Date().getTime();
                        //mSession.setSyncTime(new Date().getTime());
                        if (Contants.mFREContext != null) {
                            UploadCollectedData.formatData(getApplicationContext());
                        } else {
                            UploadCollectedData.formatData(getApplicationContext());
                        }

                    } else {
                        Log.d("current time", String.valueOf(currentTime));
                        Log.d("last time", String.valueOf(lastTime));

                    }
                }

            }, 0, 10*1000, TimeUnit.MILLISECONDS);
        } catch (Exception e) {
        }
    }
}

File: app/src/main/java/com/motrixi/datacollection/service/MotrixiService.kt
package com.motrixi.datacollection.service

import android.app.Notification
import android.app.NotificationChannel
import android.app.NotificationManager
import android.app.Service
import android.content.Context
import android.content.Intent
import android.os.Build
import android.os.IBinder
import android.support.annotation.RequiresApi
import android.util.Log
import com.motrixi.datacollection.content.Session
import com.motrixi.datacollection.utils.UploadCollectedData
import com.motrixi.datacollection.utils.UploadLogUtil
import org.apache.commons.lang3.concurrent.BasicThreadFactory
import java.util.*
import java.util.concurrent.ScheduledExecutorService
import java.util.concurrent.ScheduledThreadPoolExecutor
import java.util.concurrent.TimeUnit


/**
 * author : Jason
 *  date   : 2020/11/3 4:48 PM
 *  desc   :
 */
class MotrixiService: Service() {

    private var startUploadExecutor: ScheduledExecutorService? = null
    private val CHANNEL_ID = "motrixi_sdk_channel"
    private val CHANNEL_NAME = "motrixi_sdk_foreground_service"
    private var id = 100
    private var mSession: Session? = null
//    private val TIME_VALUE: Long = 24*60*60*1000
    private val TIME_VALUE: Long = 3*60*1000


    override fun onCreate() {
        super.onCreate()

        Log.e("oncreate","create service")
        mSession = Session(this)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            setForeground()
        }
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        /*var builder: Notification.Builder = Notification.Builder(this.applicationContext)
        var notification: Notification = builder.build()

        startForeground(100, notification)*/

        startUploading()

        return super.onStartCommand(intent, flags, startId)
    }

    override fun onBind(p0: Intent?): IBinder? {
        return null
    }


    @RequiresApi(Build.VERSION_CODES.O)
    private fun setForeground() {
        val manager = getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
        val channel = NotificationChannel(CHANNEL_ID, CHANNEL_NAME, NotificationManager.IMPORTANCE_HIGH)
        manager.createNotificationChannel(channel)
        val notification: Notification = Notification.Builder(this, CHANNEL_ID)
            //.setContentTitle("收到一条重要通知")
            //.setContentText("这是重要通知")
            //.setSmallIcon(R.mipmap.ic_launcher)
//            .setLargeIcon(BitmapFactory.decodeResource(resources, R.mipmap.ic_launcher))
            .build()
        startForeground(id, notification)
        Log.e("action","startForeground service")
    }

    /**
     * start upload the data
     */
    private fun startUploading() {

        try {
            if (startUploadExecutor == null) {
                startUploadExecutor = ScheduledThreadPoolExecutor(1,
                    BasicThreadFactory.Builder().namingPattern("start-thread").daemon(true).build())
            }
            startUploadExecutor!!.scheduleAtFixedRate(object : Runnable{
                override fun run() {
                    var currentTime = Date().time
                    var lastTime = mSession!!.syncTime
                    if (lastTime.toInt() == 0) {
                        Log.d("not confirm", "have not confirm")
                        //UploadLogUtil.uploadLogData(this@MotrixiService, "have not confirm")
                        return
                    }
                    if ((currentTime - lastTime) >= TIME_VALUE) {

                        Log.d("service", "start service")
                        UploadLogUtil.uploadLogData(this@MotrixiService, "upload data via service")
                        mSession!!.syncTime = Date().time
                        UploadCollectedData.formatData(applicationContext)
                    } else {
                        Log.d("current time", currentTime.toString())
                        Log.d("last time", lastTime.toString())

                        //UploadLogUtil.uploadLogData(this@MotrixiService, "current time$currentTime")
                    }
                }

            }, 0, 10*1000, TimeUnit.MILLISECONDS)
        } catch (e: Exception) {
        }
    }


    override fun onDestroy() {
        super.onDestroy()
    }
}